fdbclient/ManagementAPI.actor.cpp

```diff
@@ -465,6 +467,160 @@ bool isCompleteConfiguration(std::map<std::string, std::string> const& options)
 	       options.count(p + "storage_engine") == 1;
 }
 
+bool validTenantAndEncryptionAtRestMode(Optional<DatabaseConfiguration> oldConfiguration,
+                                        std::map<std::string, std::string> newConfig,
+                                        bool creating) {
+	if (creating) {
+		if (newConfig.count(encryptionAtRestModeConfKey.toString()) != 0) {
+			EncryptionAtRestMode encryptMode = EncryptionAtRestMode::fromValueRef(
+			    ValueRef(newConfig.find(encryptionAtRestModeConfKey.toString())->second));
+			if (encryptMode.mode == EncryptionAtRestMode::DISABLED ||
+			    encryptMode.mode == EncryptionAtRestMode::CLUSTER_AWARE) {
+				// In these cases the tenantMode does not matter
+				return true;
+			}
+
+			// check if the tenant mode is being set during configure new (otherwise assume tenants are disabled)
+			TenantMode tenantMode = TenantMode::DISABLED;
+			if (newConfig.count(tenantModeConfKey.toString()) != 0) {
+				tenantMode = TenantMode::fromValue(ValueRef(newConfig.find(tenantModeConfKey.toString())->second));
+			}
+			TraceEvent(SevDebug, "EncryptAndTenantModes")
+			    .detail("EncryptMode", encryptMode.toString())
+			    .detail("TenantMode", tenantMode.toString());
+			if (tenantMode != TenantMode::REQUIRED) {
+				// For domain aware encryption only the required tenant mode is currently supported
+				TraceEvent(SevWarnAlways, "InvalidEncryptAndTenantConfiguration")
+				    .detail("EncryptMode", encryptMode.toString())
+				    .detail("TenantMode", tenantMode.toString());
+				return false;
+			}
+			return true;
+		}
+	} else {
+		if (newConfig.count(tenantModeConfKey.toString()) != 0) {
+			ASSERT(oldConfiguration.present());
+			// If the tenant mode is being changed then make sure it obeys the current cluster encryption state
+			EncryptionAtRestMode encryptMode = oldConfiguration.get().encryptionAtRestMode;
+
+			if (encryptMode.mode == EncryptionAtRestMode::DISABLED) {
+				// In this case the tenantMode does not matter
+				return true;
+			}
+			TenantMode oldTenantMode = oldConfiguration.get().tenantMode;
+			TenantMode newTenantMode = TenantMode::fromValue(newConfig.find(tenantModeConfKey.toString())->second);
+			TraceEvent(SevDebug, "EncryptAndTenantModes")
+			    .detail("EncryptMode", encryptMode.toString())
+			    .detail("OldTenantMode", oldTenantMode.toString())
+			    .detail("NewTenantMode", newTenantMode.toString());
+			if (encryptMode.mode == EncryptionAtRestMode::DOMAIN_AWARE && newTenantMode != TenantMode::REQUIRED) {
```

Comment:
Is this a valid scenario? 
Based on above `if (creating)` block check, `if encryptionAtMode == domain_aware` **then** `tenant_mode == required`

Should this be an assert?

Reply:
This is when you're changing config. Its possible a cluster was initally configured correctly (domain_aware and tenant mode required) but when changing you try to change the tenant mode from required to something else. I don't think it should be an assert but rather an invalid configuration

Reply:
Got it. Makes sense.